File: clients/defold-app/playground/labs/lab04.lua
--
-- Lab 04 — Remote Interpolation.
--
-- The SAME bot rendered through four Predict instances, one per mode:
--
--   raw          the decoded snapshot verbatim: stutters at patch rate.
--   lerp         render `delay` ms in the PAST, between two real samples.
--                Never wrong, always late.
--   damped       exponential chase of the newest sample: lags by construction
--                and rounds corners.
--   extrapolate  project the trend FORWARD. Present-time, but overshoots
--                whenever the bot turns.
--
-- Port of src/client/labs/04-interp-modes/.
--
local sim = require 'playground.sim'
local app = require 'playground.app'
local Predict = require 'colyseus.predict.predict'

local kb = app.kb
local PATTERNS = { "patrol", "circle", "wander" }

--- Coefficient of variation of rendered per-frame speed — the "limp" metric.
--- Constant speed scores 0; a mode that stutters between stalls and jumps
--- scores high.
local Smoothness = {}
Smoothness.__index = Smoothness

function Smoothness.new()
  return setmetatable({ speeds = {}, head = 0, count = 0, seeded = false }, Smoothness)
end

function Smoothness:clear()
  self.speeds, self.head, self.count, self.seeded = {}, 0, 0, false
end

function Smoothness:sample(x, y, dt_ms)
  if self.seeded and dt_ms > 0 then
    local dx, dy = x - self.last_x, y - self.last_y
    self.head = self.head % 120 + 1
    self.speeds[self.head] = math.sqrt(dx * dx + dy * dy) / dt_ms * 1000.0
    if self.count < 120 then self.count = self.count + 1 end
  end
  self.last_x, self.last_y, self.seeded = x, y, true
end

--- nil when there isn't enough motion for the ratio to mean anything.
function Smoothness:cv()
  if self.count < 20 then return nil end
  local mean = 0
  for i = 1, self.count do mean = mean + self.speeds[i] end
  mean = mean / self.count
  if mean < 0.5 then return nil end
  local var = 0
  for i = 1, self.count do
    local d = self.speeds[i] - mean
    var = var + d * d
  end
  return math.sqrt(var / self.count) / mean
end

local Lab = {
  id = "04-interp-modes",
  num = 4,
  title = "Remote Interpolation",
  blurb = "lerp / damped / extrapolate / raw, side by side.",
  room_name = "lab-bots",
}
Lab.__index = Lab

function Lab.new()
  return setmetatable({ pattern = 1, bot_travel = 0 }, Lab)
end

function Lab:ready(room)
  return room.state.bots ~= nil and room.state.bots["bot1"] ~= nil
end

function Lab:mount(context, room)
  self.room = room
  self.sid = room.session_id
  self.bot = room.state.bots["bot1"]
  if self.bot == nil then return false end

  self.input = room:input()
  self.cmd = self.input.data
  self.pacer = app.Pacer.new(1000.0 / sim.TICK_HZ)

  -- One Predict per mode — same entity, same fields, different smoothing.
  local function attach(name, color_key, opts)
    local predict = Predict.for_room(room)
    predict:track(self.bot, "x", opts)
    predict:track(self.bot, "y", opts)
    return { name = name, color_key = color_key, predict = predict,
             visible = true, smooth = Smoothness.new() }
  end

  self.modes = {
    { name = "raw", color_key = "text", visible = true, smooth = Smoothness.new() },
    attach("lerp", "blue", { mode = "lerp", delay = sim.REMOTE_INTERP_MS }),
    attach("damped", "good", { mode = "damped", damping = 12 }),
    attach("extrapolate", "warn", { mode = "extrapolate", max_extrapolate = 250 }),
  }
  return true
end

function Lab:value(mode, field)
  if mode.predict == nil then return self.bot[field] end   -- raw: the instance
  return mode.predict:value(self.bot, field)
end

function Lab:set_pattern(kind)
  for i, p in ipairs(PATTERNS) do
    if p == kind then
      self.pattern = i
      self.room:send("pattern", { kind = kind })
      return
    end
  end
end

--- Start a clean measurement window once a pattern change has landed.
function Lab:reset_meters()
  for _, m in ipairs(self.modes) do m.smooth:clear() end
  self.bot_travel = 0
end

function Lab:frame(context, now, dt_ms)
  if kb.key("b") then
    self:set_pattern(PATTERNS[self.pattern % #PATTERNS + 1])
  end
  for i = 1, 4 do
    if kb.key("f" .. i) then self.modes[i].visible = not self.modes[i].visible end
  end

  local steps = self.pacer:steps(now)
  for _ = 1, steps do
    self.cmd.moveX = kb.move_x()
    self.cmd.moveY = kb.move_y()
    self.input:send()
  end
  for _, m in ipairs(self.modes) do
    if m.predict then m.predict:tick(now) end
  end

  if self.last_raw_x ~= nil and (self.bot.x ~= self.last_raw_x or self.bot.y ~= self.last_raw_y) then
    local dx, dy = self.bot.x - self.last_raw_x, self.bot.y - self.last_raw_y
    self.bot_travel = self.bot_travel + math.sqrt(dx * dx + dy * dy)
  end
  self.last_raw_x, self.last_raw_y = self.bot.x, self.bot.y

  for _, m in ipairs(self.modes) do
    m.smooth:sample(self:value(m, "x"), self:value(m, "y"), dt_ms)
  end
end

function Lab:render(gfx)
  self.room.state.players:each(function(p, key)
    gfx.square(p.x, p.y, sim.PLAYER_HALF, gfx.hue(p.hue, key == self.sid and 0.9 or 0.4))
  end)

  for i, m in ipairs(self.modes) do
    if m.visible then
      local x, y = self:value(m, "x"), self:value(m, "y")
      gfx.circle_outline(x, y, sim.BOT_RADIUS, gfx.PALETTE[m.color_key], i == 1)
      -- Stagger the labels: the four modes sit within a bot radius of each
      -- other most of the time, so a shared baseline is unreadable.
      gfx.label(x, y, m.name, gfx.PALETTE[m.color_key], 10, -14 - (i - 1) * 11)
    end
  end

  gfx.hud_section("SMOOTHNESS (speed CV)")
  for _, m in ipairs(self.modes) do
    local cv = m.smooth:cv()
    if cv == nil then
      gfx.hud_row(m.name, "--", gfx.PALETTE.text_faint)
    else
      gfx.hud_row(m.name, string.format("%.0f %%", cv * 100),
        cv < 0.15 and gfx.PALETTE.good or cv < 0.5 and gfx.PALETTE.warn or gfx.PALETTE.bad)
    end
  end
  gfx.hud_row("bot travel", string.format("%.0f u", self.bot_travel), gfx.PALETTE.text)

  gfx.hud_section("CONTROLS")
  gfx.hud_key("WASD", "drive your own square")
  gfx.hud_key("B", "bot pattern: " .. PATTERNS[self.pattern])
  for i, m in ipairs(self.modes) do
    gfx.hud_key("F" .. i, (m.visible and "hide " or "show ") .. m.name)
  end
  gfx.hud_note("Lower CV = smoother render. raw stutters at the patch rate; " ..
    "lerp is smooth but late; damped rounds corners; extrapolate is present-time " ..
    "but overshoots every turn — press B and watch it on circle.")
end

function Lab:unmount()
  for _, m in ipairs(self.modes) do
    if m.predict then m.predict:dispose() end
  end
end

function Lab:on_reconnect()
  self.pacer:reset()
  self.bot = self.room.state.bots["bot1"]
end

return Lab

File: clients/godot-app/scripts/labs/Lab04.cs
using System.Collections.Generic;
using System.Threading.Tasks;
using Colyseus;
using Colyseus.Predict;
using Godot;
using Lab = PredictProbe.LabSchema;

namespace Playground
{
    /// <summary>
    /// Lab 04 — Remote Interpolation.
    ///
    /// The SAME bot rendered through four Predict instances, one per mode:
    ///
    ///   raw          the decoded snapshot verbatim: stutters at patch rate.
    ///   lerp         render `delay` ms in the PAST, between two real samples.
    ///                Never wrong, always late.
    ///   damped       exponential chase of the newest sample: lags by
    ///                construction and rounds corners.
    ///   extrapolate  project the trend FORWARD. Present-time, but overshoots
    ///                whenever the bot turns.
    ///
    /// The strip along the bottom plots the bot's x over time: white dots are
    /// received samples, the coloured traces are what each mode rendered.
    /// Port of labs/04-interp-modes/.
    /// </summary>
    public partial class Lab04 : LabBase<Lab.BotsState>
    {
        public override string Id => "04-interp-modes";
        public override int Num => 4;
        public override string Title => "Remote Interpolation";
        public override string Blurb => "lerp / damped / extrapolate / raw, side by side.";

        private const int TraceCap = 400;
        private const double StripSpanMs = 2500;

        private static readonly string[] Patterns = { "patrol", "circle", "wander" };

        /// <summary>One render mode: its Predict (null = raw), trace, and metric.</summary>
        private class Mode
        {
            public string Name;
            public Color Color;
            public Predict Predict;      // null for "raw" — read the instance
            public bool Visible = true;
            public readonly Series Trace = new Series(TraceCap);
            public readonly Smoothness Smooth = new Smoothness();
        }

        private Lab.MoveInput _cmd;
        private InputHandle _input;
        private Pacer _pacer;
        private string _sid;
        private Lab.Bot _bot;

        private Mode[] _modes;
        private readonly Series _samples = new Series(TraceCap);
        private double _lastRawX = double.NaN, _lastRawY = double.NaN;
        private int _pattern;
        private double _botTravel;

        public override async Task<bool> Mount(App app)
        {
            Room = await Shell.JoinLab<Lab.BotsState>(app, "lab-bots",
                r => r.State.bots != null && r.State.bots.ContainsKey("bot1"));
            _sid = Room.SessionId;
            if (Room.State.bots == null || !Room.State.bots.TryGetValue("bot1", out _bot)) return false;

            _cmd = new Lab.MoveInput();
            _input = Room.Input(_cmd);
            _pacer = new Pacer(1000.0 / Sim.TickHz);

            _modes = new[]
            {
                new Mode { Name = "raw", Color = Palette.Text },
                Attach("lerp", Palette.Blue, new PredictFieldOptions
                    { Mode = PredictMode.Lerp, Delay = Sim.RemoteInterpMs }),
                Attach("damped", Palette.Good, new PredictFieldOptions
                    { Mode = PredictMode.Damped, SmoothMs = 83 }),
                Attach("extrapolate", Palette.Warn, new PredictFieldOptions
                    { Mode = PredictMode.Extrapolate, MaxExtrapolate = 250 }),
            };
            return true;
        }

        /// <summary>One Predict per mode — same entity, same fields, different smoothing.</summary>
        private Mode Attach(string name, Color color, PredictFieldOptions opts)
        {
            var predict = Predict.Get(Room);
            predict.Attach(_bot, new AttachConfig { ["x"] = opts, ["y"] = opts });
            return new Mode { Name = name, Color = color, Predict = predict };
        }

        private double Value(Mode m, string field)
            => m.Predict == null
                ? (field == "x" ? _bot.x : _bot.y)
                : m.Predict.Value(_bot, field);

        public override void Frame(App app, double now, double dtMs)
        {
            if (Kb.Key(Key.B))
            {
                _pattern = (_pattern + 1) % Patterns.Length;
                Room.Send("pattern", new Dictionary<string, string> { ["kind"] = Patterns[_pattern] });
            }
            for (int i = 0; i < _modes.Length; i++)
                if (Kb.Key(Key.F1 + i)) _modes[i].Visible = !_modes[i].Visible;

            int steps = _pacer.Steps(now);
            for (int i = 0; i < steps; i++)
            {
                _cmd.moveX = (sbyte)Kb.MoveX();
                _cmd.moveY = (sbyte)Kb.MoveY();
                _input.Send();
            }
            foreach (var m in _modes) m.Predict?.Tick(now);

            // A received sample = the decoded value changing, once per patch.
            if (_bot.x != _lastRawX || _bot.y != _lastRawY)
            {
                if (!double.IsNaN(_lastRawX))
                {
                    double bdx = _bot.x - _lastRawX, bdy = _bot.y - _lastRawY;
                    _botTravel += System.Math.Sqrt(bdx * bdx + bdy * bdy);
                }
                _lastRawX = _bot.x;
                _lastRawY = _bot.y;
                _samples.Push(Room.Clock.LastServerTime(), _bot.x);
            }

            double sNow = Room.Clock.ServerNow();
            foreach (var m in _modes)
            {
                double x = Value(m, "x"), y = Value(m, "y");
                m.Trace.Push(sNow, x);
                m.Smooth.Sample(x, y, dtMs);
            }
        }

        public override void Render(App app)
        {
            var v = app.View;
            Room.State.players.ForEach((key, p) =>
                Draw.Square(v, p.x, p.y, Sim.PlayerHalf, Palette.Hue(p.hue, key == _sid ? 0.9f : 0.4f)));

            for (int i = 0; i < _modes.Length; i++)
            {
                var m = _modes[i];
                if (!m.Visible) continue;
                double x = Value(m, "x"), y = Value(m, "y");
                Draw.CircleOutline(v, x, y, Sim.BotRadius, m.Color, dashed: i == 0);
                // Stagger the labels: the four modes sit within a bot radius of
                // each other most of the time, so a shared baseline is unreadable.
                Draw.Label(v, x, y, m.Name, m.Color, 10, -v.S(Sim.BotRadius) - 14 - i * 11);
            }

            DrawStrip(app);

            var h = app.Hud;
            h.Section("SMOOTHNESS (speed CV)");
            foreach (var m in _modes)
            {
                double cv = m.Smooth.Cv();
                if (double.IsNaN(cv)) h.Row(m.Name, "--", Palette.TextFaint);
                else h.Row(m.Name, $"{cv * 100:F0} %",
                    cv < 0.15 ? Palette.Good : cv < 0.5 ? Palette.Warn : Palette.Bad);
            }
            h.Row("samples buffered", _samples.Count.ToString(), Palette.Text);

            h.Section("CONTROLS");
            h.Key("WASD", "drive your own square");
            h.Key("B", $"bot pattern: {Patterns[_pattern]}");
            for (int i = 0; i < _modes.Length; i++)
                h.Key($"F{i + 1}", $"{(_modes[i].Visible ? "hide" : "show")} {_modes[i].Name}");
            h.Note("Lower CV = smoother render. raw stutters at the patch rate; lerp is " +
                   "smooth but late; damped rounds corners; extrapolate is present-time but " +
                   "overshoots every turn — press B and watch it on circle.");
        }

        /// <summary>The buffer timeline: received samples vs per-mode rendered traces.</summary>
        private void DrawStrip(App app)
        {
            const float h = 118, pad = 10;
            float x0 = app.Stage.Position.X + 24, x1 = app.Stage.End.X - 16;
            float y0 = app.Stage.End.Y - h - 10, y1 = y0 + h;
            if (x1 - x0 < 200) return;

            Draw.Rect(new Rect2(x0, y0, x1 - x0, h), Palette.A(Palette.Panel, 0.9f));

            double sNow = Room.Clock.ServerNow(), lastPatch = Room.Clock.LastServerTime();
            double tMin = sNow - StripSpanMs, tMax = sNow + 150;
            // Value axis: bot x spans ~[20,80]; leave headroom for overshoot.
            const double vMin = 12, vMax = 88;
            float TX(double t) => x0 + pad + (float)((t - tMin) / (tMax - tMin)) * (x1 - x0 - pad * 2);
            float TY(double val) => y1 - pad - (float)((val - vMin) / (vMax - vMin)) * (h - pad * 2 - 14);

            foreach (var (t, label) in new[] { (lastPatch, "newest patch"), (sNow, "now") })
            {
                float x = TX(t);
                if (x < x0 || x > x1) continue;
                Draw.Dashed(new Vector2(x, y0 + 4), new Vector2(x, y1 - 4),
                    Palette.A(Palette.Text, 0.25f), 1f, 3f, 3f);
                Draw.Text(x - 40, y0 + 6, 80, label, Palette.A(Palette.Text, 0.5f), 9,
                    HorizontalAlignment.Center);
            }

            _samples.ForEach((t, val) =>
            {
                if (t < tMin || t > tMax) return;
                Draw.Rect(new Rect2(TX(t) - 1.5f, TY(val) - 1.5f, 3, 3), Palette.A(Palette.Text, 0.85f));
            });

            foreach (var m in _modes)
            {
                if (!m.Visible || m.Predict == null) continue;
                bool started = false;
                Vector2 prev = default;
                m.Trace.ForEach((t, val) =>
                {
                    if (t < tMin || t > tMax) { started = false; return; }
                    var pt = new Vector2(TX(t), TY(val));
                    if (started) Draw.Line(prev, pt, Palette.A(m.Color, 0.9f), 1.2f);
                    prev = pt;
                    started = true;
                });
            }
        }

        /// <summary>Drive the bot pattern from the harness, as the B key would.</summary>
        public void SetPattern(string kind)
        {
            int i = System.Array.IndexOf(Patterns, kind);
            if (i < 0) return;
            _pattern = i;
            Room.Send("pattern", new Dictionary<string, string> { ["kind"] = kind });
        }

        /// <summary>How far the raw bot has travelled — a stationary bot scores NaN.</summary>
        public double BotTravel => _botTravel;

        /// <summary>Start a clean measurement window once the pattern has landed.</summary>
        public void ResetMeters()
        {
            foreach (var m in _modes) m.Smooth.Clear();
            _botTravel = 0;
        }

        /// <summary>Per-mode speed CV, for the acceptance harness.</summary>
        public Dictionary<string, double> SmoothnessByMode()
        {
            var r = new Dictionary<string, double>();
            foreach (var m in _modes) r[m.Name] = m.Smooth.Cv();
            return r;
        }

        /// <summary>Per-mode meter state, so a NaN can explain itself.</summary>
        public string DescribeMode(string name)
        {
            foreach (var m in _modes) if (m.Name == name) return m.Smooth.Describe();
            return "?";
        }

        public override void Unmount()
        {
            foreach (var m in _modes) m.Predict?.Dispose();
        }

        public override void OnReconnect()
        {
            _pacer.Reset();
            Room.State.bots.TryGetValue("bot1", out _bot);
        }
    }
}
